lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -1175,11 +1099,20 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
   {
     SmallVector<Substitution, 4> List;
 
+    // Define a substitution map for re-mapping interface types of
+    // the specialized function to contextual types of the caller.
+    auto SpecializedInterfaceToCallerArchetypeMap =
+        SpecializedGenericSig->getSubstitutionMap(
+            [&](SubstitutableType *type) -> Type {
+              return SpecializedInterfaceToCallerArchetypeMapping.lookup(
+                  type->getCanonicalType());
+            },
+            LookUpConformanceInSignature(*SpecializedGenericSig));
+
     SpecializedGenericSig->getSubstitutions(
         [&](SubstitutableType *type) -> Type {
           auto Replacement =
-              SpecializedInterfaceToCallerArchetypeMapping.lookup(
-                  type->getCanonicalType());
+              Type(type).subst(SpecializedInterfaceToCallerArchetypeMap);
           if (Replacement)
             return Replacement;
```

Comment:
Hmm. I guess you are right. The only way how it could fail is when a substitution is concrete, but in this case it is known already and the generic signature would contain `T == ConcreteType` for it and thus getSubstitutions will not ask for it, right? Because it only collects substitutions which are really supposed to be provided by an apply.